e2e: node: testcase to exercise device manager recovery

ffromani opened the pull request:
When kubelet recovers pods, there's no guarantee about order of recovering. This means a pod requesting devices can be recovered, and can start, before the device plugin governing the device is started.

The end result is that containers can be admitted and start running without the requested device actually available, causing runtime errors in the workload. This is u/s issue
https://github.com/kubernetes/kubernetes/issues/109595

With this commit, we add a e2e test, focusing on the most critical SNO scenario, which exercise the flow and verify the system behavior is fixed and workload pods requesting devices are rejected (not admitted) if they are recovered before the relevant device plugin.

ffromani commented:
/hold
still WIP

ffromani commented:
the PR is now completed. cc @Tal-or @swatisehgal

swatisehgal commented on test/extended/devices/recovery.go:
```diff
@@ -0,0 +1,416 @@
+}
+
+func unblockRegistrationControlOnNodeOrFail(oc *exutil.CLI, node *corev1.Node) {
+	framework.Logf("Preparing node %q for sample device plugin registration", node.Name)
+	fwk := oc.KubeFramework()
+	_, err := exutil.ExecCommandOnMachineConfigDaemon(fwk.ClientSet, oc, node, []string{
+		"sh", "-c", unblockRegistrationCommandMCD,
+	})
+	o.Expect(err).NotTo(o.HaveOccurred())
+	framework.Logf("Prepared node %q for sample device plugin registration", node.Name)
```
NIT: Log messages here need to be updated to capture that they for unblocking the device plugin registration process.

ffromani replied:
good point. Will add to the existing log.

ffromani replied:
my fault, I misread. Fixed the log entry.

swatisehgal commented:
Other than the minor NIT, the PR looks good to me. Thanks for your work on this!

ffromani commented:
addressed the review commants and functionally complete, but still on hold because further test is pending to adjust the timeouts. On cloud SNOs the timeouts are sufficient, but we want to be able to run the tests on BM, on which reboot can take long time.

ffromani commented on test/extended/devices/recovery.go:
```diff
@@ -0,0 +1,416 @@
+)
+
+const (
+	workloadCommand               = "devs=$(ls /tmp/ | egrep '^Dev-[0-9]+$') && echo devices=$devs && sleep 24h" // sleep forever in the test timescale
```
this should fail hard if no devices are present.

ffromani replied:
Should be fixed now, but requires verification.

Tal-or commented on test/extended/devices/recovery.go:
```diff
@@ -0,0 +1,416 @@
+
+		// phase1: figure out the node and prepare it
+		node, err := k8sCli.CoreV1().Nodes().Get(context.Background(), targetNode, metav1.GetOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred(), "error getting the target node %q", targetNode)
```
NIT: usually we use `ToNot` which is more readable
So we read:
Except error _to not_ have occurred other 
instead of:
Except error _not to_ have occurred

ffromani replied:
Absolutely, no nit at all! thanks for pointing this out. Fixed in the last upload

Tal-or commented on test/extended/devices/recovery.go:
```diff
@@ -0,0 +1,416 @@
+const (
+	workloadCommand               = "devs=$(ls /tmp/ | egrep '^Dev-[0-9]+$') && echo devices=$devs && sleep 24h" // sleep forever in the test timescale
+	setupRegistrationCommandMCD   = "mkdir -p /rootfs/var/lib/kubelet/device-plugins/sample && touch /rootfs/var/lib/kubelet/device-plugins/sample/registration"
+	unblockRegistrationCommandMCD = "rm /rootfs/var/lib/kubelet/device-plugins/sample/registration"
```
let's remove also the `sample` directory because it's not part of the original content of the node

Tal-or reviewed (commented):
Looking very good.
One question I've though - in case of failure for any reason, can we make sure (or at least try) to delete the registration file and `sample` dir from the node?

yanirq commented on test/extended/devices/recovery.go:
```diff
@@ -0,0 +1,420 @@
+
+	sampleDevicePluginPodDSYAML = "test/e2e/testing-manifests/sample-device-plugin/sample-device-plugin-control-registration.yaml"
+
+	targetNodeEnvVar = "DEVICE_RECOVERY_TARGET_NODE"
```
where does the actual setup of the DEVICE_RECOVERY_TARGET_NODE env var happens ? in selective tests ?  
will the test suits running for this PR actually include that or just skip the tests?

ffromani replied:
The entity (human or program) running the test is in charge of setting the environment variable, otherwise the test will always skip. Thing is, didn't find obvious way to pick a good target automatically, so having the explicit selection requirement seem the lesser evil. I'm open to suggestions though.

ffromani commented:
all comments addressed. Will unhold once the tuning of the timeouts is completed.

ffromani commented:
/retest

ffromani commented:
/retest

ffromani commented:
/test ci/prow/e2e-aws-csi 
/test ci/prow/e2e-aws-ovn-fips
/test ci/prow/e2e-gcp-ovn-etcd-scaling

ffromani commented:
/hold cancel

fixed the timeouts, further refinements are non-blocking

ffromani commented:
/retest

swatisehgal reviewed (commented):
/lgtm

ffromani commented:
so, running this test will cause `openshift-tests` to emit
```
[sig-arch] unknown image: registry.k8s.io/e2e-test-images/sample-device-plugin:1.5 (container/sample-device-plugin reason/Pulled image/registry.k8s.io/e2e-test-images/sample-device-plugin:1.5)
Writing JUnit report to /home/fromani/tmp/junit/junit_e2e__20230510-143327.xml

Suite run returned error: failed because an invariant was violated, 1 pass, 0 skip (5m24s)

error: failed because an invariant was violated, 1 pass, 0 skip (5m24s)
```

Which is because the added test consumes `http://registry.k8s.io/e2e-test-images/sample-device-plugin:1.5`.
I'm going to investigate further why this image doesn't have a `ImageID` u/s I can use in `test/utils/image` (per steps in `https://github.com/openshift/origin/blob/master/test/extended/util/image/README.md`). Likely this is because the image is used by `e2e_node` - not `e2e` tests. But I'll doublecheck u/s

ffromani commented:
/retest

PR had no changes and these lanes passed before

ffromani commented:
/retest

ffromani commented:
after almost 1 year, I think this is not gonna make it. Too bad.
